### ISSUE TITLE
TypeList is phantom roled

### ISSUE THREAD

**Issue by RyanGlScott** [Lysxia/type-map#7]
`TypeList` (from `Data.TypeMap.List`) is quite susceptible to being `coerce`d to unsafe things. Here is a minimal example of what I mean:

```haskell
#!/usr/bin/env cabal
{- cabal:
build-depends: base, type-map
-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeApplications #-}
module Main where

import Data.Coerce
import Data.TypeMap.List

e :: TypeList '[ '("a", Int) ]
e = maxBound <| empty

e' :: TypeList '[ '("a", Either Char Char) ]
e' = coerce e

main :: IO ()
main = print $ index @"a" e'
```
```
$ cabal new-run Bug2.hs
Resolving dependencies...
Build profile: -w ghc-8.6.3 -O1
<elided>
Segmentation fault (core dumped)
```

`TypeList` should probably have a nominal role on its first argument (i.e., `type role TypeList nominal`) to prevent this sort of thing. I haven't checked, but it's possible that other data types in this library would need similar treatment.

**Comment by Lysxia** [comment]
Uh oh. Good catch, thanks!

**Comment by Lysxia** [comment]
Just to check, if I have one parameterized type defined in terms of another

```haskell
data Foo a = Foo (Bar a)
newtype Bar a = Bar a
```

Does  `type role Bar nominal` imply `type role Foo nominal`, so I only need the former annotation? (That situation occurs inside `TypeMap`.)

**Comment by RyanGlScott** [comment]
Indeed it does:

```
$ ghci -XRoleAnnotations
GHCi, version 8.6.3: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
λ> data Foo a = Foo (Bar a); newtype Bar a = Bar a; type role Bar nominal
λ> :i Foo
type role Foo nominal
data Foo a = Foo (Bar a)
        -- Defined at <interactive>:1:1
```

**Comment by Lysxia** [comment]
Great! Fixed in 7b6c662.
